torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -304,63 +242,84 @@ static THPObjectPtr wrapTuple(PyObject* index) {
   return res;
 }
 
+// NOTE: Here is the dispatch structure for `THPVariable_getitem`:
+//
+// 1. Python 1-D getter calls C++ `at::indexing::get_item` after
+// converting Python index to C++ TensorIndex.
+//
+// 2. Python N-D getter calls C++ `at::indexing::handleDimInMultiDimIndexing`
+// for each dim, after converting Python index to C++ TensorIndex. If advanced
+// indexing is needed, it calls C++ `at::indexing::dispatch_index`.
 PyObject* THPVariable_getitem(PyObject* self, PyObject* index) {
   HANDLE_TH_ERRORS
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
   OptionalDeviceGuard device_guard(device_of(self_));
 
-  // handle simple types: integers, slices, ellipsis
+  // handle simple types: none, ellipsis
   if (index == Py_None) {
-    return wrap(self_.unsqueeze(0));
+    return THPVariable_Wrap(
+      at::indexing::get_item(self_, {at::indexing::TensorIndex(at::indexing::None)}));
   } else if (index == Py_Ellipsis) {
-    return wrap(at::alias(self_));
-  } else if (THPUtils_checkLong(index)) {
-    return wrap(applySelect(self_, 0, index));
+    return THPVariable_Wrap(
+      at::indexing::get_item(self_, {at::indexing::TensorIndex(at::indexing::Ellipsis)}));
+  }
+
+  bool is_tracing = torch::jit::tracer::isTracing();
+
+  // handle simple types: integers, slices, bool
+  if (THPUtils_checkLong(index)) {
+    if (is_tracing && THPVariable_Check(index)) {
+      recordSelectTrace(THPVariable_Unpack(index));
+    }
+    return THPVariable_Wrap(
+      at::indexing::get_item(self_, {at::indexing::TensorIndex(THPUtils_unpackLong(index))}));
   } else if (PySlice_Check(index)) {
-    return wrap(applySlice(self_, 0, index, true));
+    Py_ssize_t start, stop, step;
+    checkUnpackSlice(index, &start, &stop, &step);
+    if (is_tracing) {
+      recordSliceTrace(index);
+    }
+    return THPVariable_Wrap(
+      at::indexing::get_item(
+        self_, {at::indexing::TensorIndex({start, stop, step})}, /*is_tracing_and_1d_slice_or_Nd=*/is_tracing));
```

Comment:
"Why is this the only one that gets `is_tracing`" (Ans: because that's what it's called) (Q: But why?!)